packages/kbn-telemetry-tools/GUIDELINE.md

```diff
@@ -0,0 +1,211 @@
+    - [2. Type the `fetch` function](#2-type-the-fetch-function)
+    - [3. Add a `schema` field](#3-add-a-schema-field)
+    - [4. Run the telemetry check](#4-run-the-telemetry-check)
+    - [5. Update the stored json files](#5-update-the-stored-json-files)
+  - [Updating the collector schema](#updating-the-collector-schema)
+  - [Writing the schema](#writing-the-schema)
+    - [Basics](#basics)
+    - [Allowed types](#allowed-types)
+    - [Dealing with arrays](#dealing-with-arrays)
+  - [Schema Restrictions](#schema-restrictions)
+    - [Root of schema can only be an object](#root-of-schema-can-only-be-an-object)
+
+
+## Adding schema to your collector
+
+To add a `schema` to the collector, follow these steps until the telemetry check passes.
+To check the next step needed simply run the telemetry check with the path of your collector:
+
+```
+node scripts/telemetry_check.js --path=<relative_path_to_collector>.ts
+```
+
+### 1. Update the telemetryrc file
+
+Make sure your collector is not excluded in the `telemetryrc.json` files (located at the root of the kibana project, and another on in the `x-pack` dir).
+
+```s
+[
+  {
+    ...
+    "exclude": [
+      "<path_to_my_collector>"
+    ]
+  }
+]
+```
+
+Note that the check will fail if the collector in --path is excluded.
+
+### 2. Type the `fetch` function
+1. Make sure the return of the `fetch` function is typed.
+
+The function `makeUsageCollector` accepts a generic type parameter of the returned type of the `fetch` function. 
+
+```
+interface Usage {
+  someStat: number;
+}
+
+usageCollection.makeUsageCollector<Usage>({
+  fetch: async () => {
+    return {
+      someStat: 3,
+    }
+  },
+  ...
+})
+```
+
+The generic type passed to `makeUsageCollector` will automatically unwrap the `Promise` to check for the resolved type.
+
+### 3. Add a `schema` field
+
+Add a `schema` field to your collector. After passing the return type of the fetch function to the `makeUsageCollector` generic parameter. It will automaticallly figure out the correct type of the schema based on that provided type.
+
+
+```
+interface Usage {
+  someStat: number;
+}
+
+usageCollection.makeUsageCollector<Usage>({
+  schema: {
+    someStat: {
+      type: 'number'
+    }
+  },
+  ...
+})
+```
+
+For full details on writing the `schema` object, check the [Writing the schema](#writing-the-schema) section.
+
+### 4. Run the telemetry check
+
+To make sure your changes pass the telemetry check you can run the following:
+
+```
+node scripts/telemetry_check.js --ignore-stored-json --path=<relative_path_to_collector>.ts
+```
+
+### 5. Update the stored json files
+
+The `--fix` flag will automatically update the persisted json files used by the telemetry team.
+
+```
+node scripts/telemetry_check.js --fix
+```
+
+Note that any updates to the stored json files will require a review by the kibana-telemetry team to help us update the telemetry cluster mappings and ensure your changes adhere to our best practices.
```

Comment:
This step is still unclear to me:

Should other devs edit the JSON files with the merged schemas or should we do so after FF so we know in one unique PR all the changes coming up in the new version? 🤔

Reply:
Already discussed in the PR that adds the `--fix` functionality. Resolving this question here 👍